codepropertygraph/src/main/scala/io/shiftleft/codepropertygraph/cpgloading/ProtoCpgLoader.scala

```diff
@@ -0,0 +1,211 @@
+package io.shiftleft.codepropertygraph.cpgloading
+
+import java.io.{File, FileInputStream, IOException, InputStream}
+import java.nio.file.{Files, Path}
+import java.util.Optional
+import java.util.stream.Collectors
+
+import io.shiftleft.codepropertygraph.Cpg
+import io.shiftleft.proto
+import io.shiftleft.proto.cpg.Cpg.{CpgOverlay, CpgStruct}
+import org.apache.commons.io.FileUtils
+import org.apache.logging.log4j.LogManager
+
+import scala.collection.JavaConverters._
+
+class ProtoCpgLoader {}
+
+/**
+  * Warning - this is an internal class. Please use `CpgLoader` if you
+  * require a stable API.
+  * */
+object ProtoCpgLoader {
+
+  private val logger = LogManager.getLogger(classOf[ProtoCpgLoader])
+
+  /**
+    * Load code property graph from zip archive. The zip archive
+    * contains proto files. The default loader configuration is used.
+    * @param filename filename of the CPG archive
+    * */
+  def loadFromProtoZip(filename: String): Cpg =
+    loadFromProtoZip(filename, CpgLoaderConfig.default)
+
+  /**
+    * Load code property graph from zip archive. The zip archive
+    * contains proto files.
+    * @param filename filename of the CPG archive
+    * @param config loader configuration
+    * */
+  def loadFromProtoZip(filename: String, config: CpgLoaderConfig): Cpg = {
+    extractArchiveAndExecuteOnDir[Cpg](filename, "cpg2sp_proto", { tempDirPathName =>
+      loadFromProtobufDirectory(tempDirPathName, config)
+    })
+  }
+
+  /**
+    * Load overlays from archive of overlays stored at `filename`
+    * @param filename the filename of the archive
+    * */
+  def loadOverlays(filename: String): List[proto.cpg.Cpg.CpgOverlay] =
+    extractArchiveAndExecuteOnDir(filename, "cpg2sp_proto_overlay", { tempDirPathName =>
+      loadOverlaysFromProtobufDirectory(tempDirPathName)
+    })
+
+  /**
+    * Load overlays from directory containing overlays in proto format.
+    * @inputDirectory directory containing overlays
+    * */
+  def loadOverlaysFromProtobufDirectory(inputDirectory: String): List[proto.cpg.Cpg.CpgOverlay] = {
+
+    def comparator(file1: String, file2: String): Int = {
+      val file1Split = file1.split("_")
+      val file2Split = file2.split("_")
+      if (file1Split.length < 2 || file2Split.length < 2) {
+        file1.compareTo(file2)
+      } else {
+        Integer.parseInt(file1Split(0)) - Integer.parseInt(file2Split(0))
+      }
+    }
+
+    val filesInDirectory = getFileNamesInDirectory(new File(inputDirectory))
+    // Note: sorting is necessary here because we need to load overlays
+    // in the right order. Sorting requires us to eagerly accumulate
+    // all filenames here. For the directory of overlays, this is not
+    // problematic though, since the number of overlay files in rather small.
+    // If it does ever grow, we may need to rethink this
+    filesInDirectory.sorted(comparator)
```

Comment:
```suggestion
    filesInDirectory = filesInDirectory.sorted(comparator)
```
Pretty sure you don't have any guarantee that the stream isn't returning the sorted copy instead of in-place sorting?